File: Freedom/新浪微博/基本类库/基本类/SinaTabBarController.m
//  XFTabBarViewController.m
//  Freedom
//  Created by Fay on 15/9/13.
#import "SinaTabBarController.h"
#import "XFHomeViewController.h"
#import "XFMessageViewController.h"
#import "XFProfileViewController.h"
#import "XFDiscoverViewController.h"
#import "XFNavigationController.h"
#import "FXBlurView.h"
#import "XFComposeViewController.h"
#import "XFOAuthController.h"
#import "XFAccount.h"
#import "XFAccountTool.h"
#import "XFNewFeatureController.h"
@class XFTabBar;
@protocol XFTabBarDelegate <UITabBarDelegate>
@optional
- (void)tabBarDidClickPlusButton:(XFTabBar *)tabBar;
@end
@interface XFTabBar : UITabBar
@property(nonatomic,weak)id <XFTabBarDelegate> delegate;
@end
@interface XFTabBar ()
@property (nonatomic, weak) UIButton *plusBtn;
@end
@implementation XFTabBar
@dynamic delegate;
-(id)initWithFrame:(CGRect)frame {
    
    self = [super initWithFrame:frame];
    
    if (self) {
        
        // 添加一个按钮到tabbar中
        UIButton *plusBtn = [[UIButton alloc] init];
        [plusBtn setBackgroundImage:[UIImage imageNamed:@"tabbar_compose_button"] forState:UIControlStateNormal];
        [plusBtn setBackgroundImage:[UIImage imageNamed:@"tabbar_compose_button_highlighted"] forState:UIControlStateHighlighted];
        [plusBtn setImage:[UIImage imageNamed:@"tabbar_compose_icon_add"] forState:UIControlStateNormal];
        [plusBtn setImage:[UIImage imageNamed:@"tabbar_compose_icon_add_highlighted"] forState:UIControlStateHighlighted];
        plusBtn.frameSize = plusBtn.currentBackgroundImage.size;
        [plusBtn addTarget:self action:@selector(plusClick) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:plusBtn];
        self.plusBtn = plusBtn;
        
    }
    
    
    return self;
    
}
-(void)plusClick {
    
    //通知代理
    
    if ([self.delegate respondsToSelector:@selector(tabBarDidClickPlusButton:)]) {
        [self.delegate tabBarDidClickPlusButton:self];
    }
    
}
-(void)layoutSubviews {
    
    [super layoutSubviews];
    
    
    //设置加号的位置
    
    self.plusBtn.center = CGPointMake(APPW *0.5, 50 * 0.5);
    
    //设置其他tabbarButton的位置和尺寸
    CGFloat tabBarButtonW  = APPW / 5;
    CGFloat tabbarButtonIndex = 0;
    
    for (UIView *child in self.subviews) {
        Class class = NSClassFromString(@"UITabBarButton");
        if ([child isKindOfClass:class]) {
            child.frameWidth = tabBarButtonW;
            child.frameX = tabbarButtonIndex *tabBarButtonW;
            
            //增加索引
            tabbarButtonIndex ++;
            
            if (tabbarButtonIndex == 2) {
                tabbarButtonIndex ++;
            }
        }
    }
    
    
}
@end
// 版权属于原作者
// http://code4app.com (cn) http://code4app.net (en)
// 发布代码于最专业的源码分享网站: Code4App.com
@interface SinaTabBarController ()<XFTabBarDelegate>
@property (nonatomic,weak)UIButton *plus;
@property (nonatomic,weak)FXBlurView *blurView;
@property (nonatomic,weak)UIImageView *text;
@property (nonatomic,weak)UIImageView *ablum;
@property (nonatomic,weak)UIImageView *camera;
@property (nonatomic,weak)UIImageView *sign;
@property (nonatomic,weak)UIImageView *comment;
@property (nonatomic,weak)UIImageView *more;
@end
@implementation SinaTabBarController
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
       // 设置子控制器
    XFHomeViewController *home = [[XFHomeViewController alloc]init];
    [self addChildViewController:home title:@"首页" image:@"tabbar_home" selImage:@"tabbar_home_selected"];
    
    
    XFMessageViewController *messageCenter = [[XFMessageViewController alloc] init];
    [self addChildViewController:messageCenter title:@"消息" image:@"tabbar_message_center" selImage:@"tabbar_message_center_selected"];
    
    
    XFDiscoverViewController *discover = [[XFDiscoverViewController alloc] init];
    [self addChildViewController:discover title:@"发现" image:@"tabbar_discoverS" selImage:@"tabbar_discover_selectedS"];
    
    
    XFProfileViewController *profile = [[XFProfileViewController alloc] init];
    [self addChildViewController:profile title:@"我" image:@"tabbar_profile" selImage:@"tabbar_profile_selected"];
    
    
    //更换系统自带的tabbar
    XFTabBar *tab = [[XFTabBar alloc]init];
    tab.delegate = self;
    [self setValue:tab forKey:@"tabBar"];
    XFAccount *account = [XFAccountTool account];
    //设置根控制器
    if (account) {//第三方登录状态
        // 切换窗口的根控制器
        NSString *key = @"version";
        // 上一次的使用版本（存储在沙盒中的版本号）
        NSString *lastVersion = [[NSUserDefaults standardUserDefaults] objectForKey:key];
        // 当前软件的版本号（从Info.plist中获得）
        NSString *currentVersion = [NSBundle mainBundle].infoDictionary[@"CFBundleShortVersionString"];
        if ([currentVersion isEqualToString:lastVersion]) { // 版本号相同：这次打开和上次打开的是同一个版本
            return;
        } else { // 这次打开的版本和上一次不一样，显示新特性
            [[UIApplication sharedApplication].delegate.window.rootViewController presentViewController:[[XFNewFeatureController alloc] init] animated:YES completion:^{
            }];
            // 将当前的版本号存进沙盒
            [[NSUserDefaults standardUserDefaults] setObject:currentVersion forKey:key];
            [[NSUserDefaults standardUserDefaults] synchronize];
        }
        
    } else{
#pragma mark 假数据
        NSDictionary *acont = @{@"access_token":@"2.00IjAFKG0H7CVc7e020836340bdlSS",
                                @"expires_in":@"2636676",
                                @"remind_in":@"2636676",
                                @"uid":@"5645754790"};
        account = [XFAccount accountWithDict:acont];
        //储存账号信息
        [XFAccountTool saveAccount:account];
//        [NSKeyedArchiver archiveRootObject:account toFile:[[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject] stringByAppendingPathComponent:@"account.data"]];
        [self presentViewController:[[XFOAuthController alloc]init] animated:YES completion:^{
        }];
//        [UIApplication sharedApplication].delegate.window.rootViewController = [[XFOAuthController alloc]init];
    }
    
}
//添加子控制器
-(void)addChildViewController:(UIViewController *)childVc  title:(NSString *)title image:(NSString *)image selImage:(NSString *)selImage {
    
    //设置子控制器的TabBarButton属性
    childVc.title = title;
    childVc.tabBarItem.image = [UIImage imageNamed:image];
    childVc.tabBarItem.selectedImage = [[UIImage imageNamed:selImage] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    
    NSMutableDictionary *AttrDic = [NSMutableDictionary dictionary];
    
    AttrDic[NSForegroundColorAttributeName] = [UIColor grayColor];
    
    [childVc.tabBarItem setTitleTextAttributes:AttrDic forState:UIControlStateNormal];
    
    NSMutableDictionary *selAttr = [NSMutableDictionary dictionary];
    
    selAttr[NSForegroundColorAttributeName] = [UIColor orangeColor];
    
    [childVc.tabBarItem setTitleTextAttributes:selAttr forState:UIControlStateSelected];
    
       //让子控制器包装一个导航控制器
    XFNavigationController *nav = [[XFNavigationController alloc]initWithRootViewController:childVc];
    
   
    [self addChildViewController:nav];
    
}
-(void)tabBarDidClickPlusButton:(XFTabBar *)tabBar {
    
    FXBlurView *blurView = [[FXBlurView alloc]initWithFrame:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height)];
    blurView.tintColor = [UIColor clearColor];
    self.blurView = blurView;
    
    [self.view addSubview:blurView];
    
    UIImageView *compose = [[UIImageView alloc]init];
    [compose setImage:[UIImage imageNamed:@"compose_slogan"]];
    compose.frame = CGRectMake(0, 100, self.view.frame.size.width, 48);
    compose.contentMode = UIViewContentModeCenter;
    [blurView addSubview:compose];
    
    
    UIView *bottom = [[UIView alloc]init];
    
    bottom.frame = CGRectMake(0, self.view.bounds.size.height - 44, self.view.bounds.size.height, 44);
    
    bottom.backgroundColor = [UIColor whiteColor];
    
    //bottom.contentMode = UIViewContentModeCenter;
    
    UIButton *plus = [UIButton buttonWithType:UIButtonTypeCustom];
 
    plus.frame = CGRectMake((self.view.bounds.size.width - 25) * 0.5, 8, 25, 25);
    
    [plus setImage:[UIImage imageNamed:@"tabbar_compose_background_icon_add"] forState:UIControlStateNormal];
    
    [bottom addSubview:plus];
    
    [UIView animateWithDuration:0.2 animations:^{
        
        plus.transform = CGAffineTransformMakeRotation(M_PI_4);
        self.plus = plus;
    }];
    
    [plus addTarget:self action:@selector(closeClick) forControlEvents:UIControlEventTouchUpInside];
    
    
    [blurView addSubview:bottom];
    
    UIImageView *text = [self btnAnimateWithFrame:CGRectMake(31, 500, 71, 100) imageName:@"tabbar_compose_idea" text:@"文字" animateFrame:CGRectMake(31, 280, 71, 100) delay:0.0];
    [self setAction:text action:@selector(compose)];
    self.text = text;
    
    UIImageView *ablum = [self btnAnimateWithFrame:CGRectMake(152, 500, 71, 100) imageName:@"tabbar_compose_photo" text:@"相册" animateFrame:CGRectMake(152, 280, 71, 100) delay:0.1];
    self.ablum = ablum;
    
    UIImageView *camera = [self btnAnimateWithFrame:CGRectMake(273, 500, 71, 100) imageName:@"tabbar_compose_camera" text:@"摄影" animateFrame:CGRectMake(273, 280, 71, 100) delay:0.15];
    self.camera = camera;
    
    UIImageView *sign = [self btnAnimateWithFrame:CGRectMake(31, 700, 71, 100) imageName:@"tabbar_compose_lbs" text:@"签到" animateFrame:CGRectMake(31, 410, 71, 100) delay:0.2];
    self.sign = sign;
    
    
    UIImageView *comment = [self btnAnimateWithFrame:CGRectMake(152, 700, 71, 100) imageName:@"tabbar_compose_review" text:@"评论" animateFrame:CGRectMake(152, 410, 71, 100) delay:0.25];
    self.comment = comment;
    
    UIImageView *more = [self btnAnimateWithFrame:CGRectMake(273, 700, 71, 100) imageName:@"tabbar_compose_more" text:@"更多" animateFrame:CGRectMake(273, 410, 71, 100) delay:0.3];
    self.more = more;
    
    
}
//按钮出来动画
-(UIImageView *)btnAnimateWithFrame:(CGRect)frame imageName:(NSString *)imageName text:(NSString *)text animateFrame:(CGRect)aniFrame delay:(CGFloat)delay {
    
    UIImageView *btnContainer = [[UIImageView alloc]init];
    
    btnContainer.frame  = frame;
    
    UIImageView *image = [[UIImageView alloc]initWithImage:[UIImage imageNamed:imageName]];
    
    [btnContainer addSubview:image];
    
    UILabel *word = [[UILabel alloc]initWithFrame:CGRectMake(0, 75, 71, 25)];
    [word setText:text];
    [word setTextAlignment:NSTextAlignmentCenter];
    [word setFont:[UIFont systemFontOfSize:15]];
    [word setTextColor:[UIColor grayColor]];
    
    [btnContainer addSubview:word];
    
    [self.blurView addSubview:btnContainer];
    
    [UIView animateWithDuration:0.5 delay:delay usingSpringWithDamping:0.6 initialSpringVelocity:0.05 options:UIViewAnimationOptionAllowUserInteraction animations:^{
    
        btnContainer.frame  = aniFrame;
    
    } completion:^(BOOL finished) {
        
    }];
    
    return btnContainer;
}
//设置按钮方法
-(void)setAction:(UIImageView *)imageView action:(SEL)action{
    
    UITapGestureRecognizer *gesture = [[UITapGestureRecognizer alloc]initWithTarget:self action:action];
    imageView.userInteractionEnabled = YES;
    [imageView addGestureRecognizer:gesture];
    
}
//发文字微博
-(void)compose {
   
    [self closeClick];
    
    XFComposeViewController *compose = [[XFComposeViewController alloc]init];
    
    XFNavigationController *nav = [[XFNavigationController alloc]initWithRootViewController:compose];
    
    [self presentViewController:nav animated:YES completion:nil];
    
    
    
    
}
//关闭动画
-(void)btnCloseAnimateWithFrame:(CGRect)rect delay:(CGFloat)delay btnView:(UIImageView *)btnView{
    
    
    [UIView animateWithDuration:0.3 delay:delay usingSpringWithDamping:0.6 initialSpringVelocity:0.05 options:UIViewAnimationOptionAllowUserInteraction animations:^{
        
        btnView.frame  = rect;
        
    } completion:^(BOOL finished) {
        
    }];
 
}
//关闭按钮
-(void)closeClick {
    
    [UIView animateWithDuration:0.6 animations:^{
        
        self.plus.transform = CGAffineTransformMakeRotation(-M_PI_2);
        [self btnCloseAnimateWithFrame:CGRectMake(273, 700, 71, 100) delay:0.1 btnView:self.more];
        [self btnCloseAnimateWithFrame:CGRectMake(152, 700, 71, 100) delay:0.15 btnView:self.comment];
        [self btnCloseAnimateWithFrame:CGRectMake(31, 700, 71, 100) delay:0.2 btnView:self.sign];
        [self btnCloseAnimateWithFrame:CGRectMake(273, 700, 71, 100) delay:0.25 btnView:self.camera];
        [self btnCloseAnimateWithFrame:CGRectMake(152, 700, 71, 100) delay:0.3 btnView:self.ablum];
        [self btnCloseAnimateWithFrame:CGRectMake(31, 700, 71, 100) delay:0.35 btnView:self.text];
    } completion:^(BOOL finished) {
        
        [self.text removeFromSuperview];
        [self.ablum removeFromSuperview];
        [self.camera removeFromSuperview];
        [self.sign removeFromSuperview];
        [self.comment removeFromSuperview];
        [self.more removeFromSuperview];
        [self.blurView removeFromSuperview];
    }];
    
}
@end
// 版权属于原作者
// http://code4app.com (cn) http://code4app.net (en)
// 发布代码于最专业的源码分享网站: Code4App.com

File: Freedom/新浪微博/基本类库/SinaTabBarController.swift
//
//  SinaTabBarController.swift
//  Freedom
//
//  Created by Super on 6/28/18.
//  Copyright © 2018 薛超. All rights reserved.
import UIKit
import XExtension
import XCarryOn
class SinaTabBar: UITabBar {
    var centerButton = UIButton()
    override init(frame: CGRect) {
        super.init(frame: frame)
        centerButton.setBackgroundImage(UIImage(named: "tabbar_compose_button"), for: .normal)
        centerButton.setImage(UIImage(named: "tabbar_compose_icon_add"), for: .normal)
        let size = (centerButton.currentBackgroundImage?.size)!
        centerButton.frame = CGRect(x: centerButton.frameX, y: centerButton.frameY, width: size.width, height: size.height)
        centerButton.center = CGPoint(x: APPW * 0.5, y: 50 * 0.5)
        addSubview(centerButton)
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override func layoutSubviews() {
        super.layoutSubviews()
//        centerButton.center = CGPoint(x: bounds.size.width * 0.5, y: 25)
        let tabbarButtonW: CGFloat = bounds.size.width / 5
        var buttonIndex: CGFloat = 0
        for view in subviews {
            if view.isKind(of: NSClassFromString("UITabBarButton")! ) {
                view.frame = CGRect(x: buttonIndex * tabbarButtonW,y:view.frame.origin.y, width: tabbarButtonW, height:view.bounds.size.height)
                buttonIndex += 1
                if buttonIndex == 2 {
                    buttonIndex += 1
                }
            }
        }
    }
}
//按钮出来动画
class SinaTabBarController: BaseTabBarViewController {
    var plus: UIButton = UIButton(frame: CGRect(x: (APPW - 25) * 0.5, y: 8, width: 25, height: 25))
    var blurView: UIView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH))
    var text: UIImageView = UIImageView(frame: CGRect(x: 31, y: 500, width: 71, height: 100))
    var ablum: UIImageView = UIImageView(frame: CGRect(x: 152, y: 500, width: 71, height: 100))
    var camera: UIImageView = UIImageView(frame: CGRect(x: 273, y: 500, width: 71, height: 100))
    var sign: UIImageView = UIImageView(frame: CGRect(x: 31, y: 700, width: 71, height: 100))
    var comment: UIImageView = UIImageView(frame: CGRect(x: 152, y: 700, width: 71, height: 100))
    var more: UIImageView = UIImageView(frame: CGRect(x: 273, y: 700, width: 71, height: 100))
    override func viewDidLoad() {
        super.viewDidLoad()
        var i=0
        for vc in self.children{
            vc.tabBarItem.setTitleTextAttributes([NSAttributedString.Key.foregroundColor : UIColor.red,NSAttributedString.Key.font : UIFont.boldSystemFont(ofSize: 12)], for: .selected)
            vc.tabBarItem.image = vc.tabBarItem.image?.withRenderingMode(.alwaysOriginal)
            vc.tabBarItem.selectedImage = vc.tabBarItem.selectedImage?.withRenderingMode(.alwaysOriginal)
            vc.tabBarItem.tag = i;
            i += 1
        }
        let tabbar = SinaTabBar()
        self.setValue(tabbar, forKeyPath: "tabBar")
        tabbar.centerButton.addTarget(self, action:#selector(self.centerClicked), for: .touchUpInside)
        initUI()
        configAccount()
    }
    private func initUI(){
        func configBtn(_ img:UIImageView,imageName:String,text:String,tag:Int){
            let image = UIImageView(image: UIImage(named: imageName ))
            let word = UILabel(frame: CGRect(x: 0, y: 75, width: 71, height: 25))
            word.text = text
            word.textAlignment = .center
            word.font = UIFont.systemFont(ofSize: 15)
            word.textColor = UIColor.gray
            img.addSubview(image)
            img.addSubview(word)
            let gesture = UITapGestureRecognizer(target: self, action: #selector(self.compose(gesture:)))
            img.isUserInteractionEnabled = true
            img.addGestureRecognizer(gesture)
            img.tag = tag
            blurView.addSubview(img)
        }
        configBtn(text, imageName: "tabbar_compose_idea", text: "文字",tag:1)
        configBtn(ablum, imageName: "tabbar_compose_photo", text: "相册",tag:2)
        configBtn(camera, imageName: "tabbar_compose_camera", text: "摄影",tag:3)
        configBtn(sign, imageName: "tabbar_compose_lbs", text: "签到",tag:4)
        configBtn(comment, imageName: "tabbar_compose_review", text: "评论",tag:5)
        configBtn(more, imageName: "tabbar_compose_more", text: "更多",tag:6)
        blurView.tintColor = UIColor.clear
        let compose = UIImageView(frame: CGRect(x: 0, y: 100, width: view.frame.size.width, height: 48))
        compose.image = UIImage(named: "compose_slogan")
        plus.setImage(UIImage(named: "tabbar_compose_background_icon_add"), for: .normal)
        plus.addTarget(self, action: #selector(self.closeClick), for: .touchUpInside)
        let bottom = UIView(frame: CGRect(x: 0, y: view.bounds.size.height - 44, width: view.bounds.size.height, height: 44))
        bottom.backgroundColor = UIColor.white
        bottom.addSubview(plus)
        blurView.addSubview(compose)
        blurView.addSubview(bottom)
        blurView.isHidden = true
        view.addSubview(blurView)
    }
    private func configAccount(){
        var account = SinaAccount.account()
        //设置根控制器
        if account != nil {
            //第三方登录状态
            let key = "version"
            let lastVersion = UserDefaults.standard.object(forKey: key) as? String
            let currentVersion = Bundle.main.infoDictionary!["CFBundleShortVersionString"] as? String
            if (currentVersion == lastVersion) {
                // 版本号相同：这次打开和上次打开的是同一个版本
                return
            } else {
                // 这次打开的版本和上一次不一样，显示新特性
                print("推出新特性的控制器")
                UserDefaults.standard.set(currentVersion, forKey: key)
                UserDefaults.standard.synchronize()
            }
        } else {
            let acont = ["access_token": "2.00IjAFKG0H7CVc7e020836340bdlSS", "expires_in": "2636676", "remind_in": "2636676", "uid": "5645754790"]
            account = SinaAccount(dict: acont)
            //储存账号信息
            SinaAccount.save(account)
            present(SinaAuthController(), animated: true) {
            }
        }
    }
    //发文字微博
    @objc private func compose(gesture:UITapGestureRecognizer) {
        closeClick()
        var vc:SinaBaseViewController?
        let view = gesture.view!
        switch view.tag {
        case 1:vc = SinaComposeViewController()
        case 2:vc = SinaComposeViewController()
        case 3:vc = SinaComposeViewController()
        case 4:vc = SinaComposeViewController()
        case 5:vc = SinaComposeViewController()
        default:vc = SinaComposeViewController()
        }
        let nav = SinaNavigationController(rootViewController: vc!)
        present(nav, animated: true)
    }
    //打开按钮
    @objc private func centerClicked(){
        func btnOpenAnimate(withFrame rect: CGRect,delay: CGFloat, btnView: UIImageView){
            UIView.animate(withDuration: 0.5, delay: TimeInterval(delay), usingSpringWithDamping: 0.6, initialSpringVelocity: 0.05, options: .allowUserInteraction, animations: {
                btnView.frame = rect
            }) { finished in
            }
        }
        btnOpenAnimate(withFrame: CGRect(x: 31, y: 280, width: 71, height: 100), delay: 0.0, btnView: text)
        btnOpenAnimate(withFrame: CGRect(x: 152, y: 280, width: 71, height: 100), delay: 0.1, btnView: ablum)
        btnOpenAnimate(withFrame: CGRect(x: 273, y: 280, width: 71, height: 100), delay: 0.15, btnView: camera)
        btnOpenAnimate(withFrame: CGRect(x: 31, y: 410, width: 71, height: 100), delay: 0.2, btnView: sign)
        btnOpenAnimate(withFrame: CGRect(x: 152, y: 410, width: 71, height: 100), delay: 0.25, btnView: comment)
        btnOpenAnimate(withFrame: CGRect(x: 273, y: 410, width: 71, height: 100), delay: 0.3, btnView: more)
        UIView.animate(withDuration: 0.2, animations: {
            self.plus.transform = CGAffineTransform(rotationAngle: .pi/4)
        })
        blurView.isHidden = false
    }
    //关闭按钮
    @objc private func closeClick() {
        //关闭动画
        func btnCloseAnimate(withFrame rect: CGRect, delay: CGFloat, btnView: UIImageView?) {
            UIView.animate(withDuration: 0.3, delay: TimeInterval(delay), usingSpringWithDamping: 0.6, initialSpringVelocity: 0.05, options: .allowUserInteraction, animations: {
                btnView?.frame = rect
            }) { finished in
            }
        }
        UIView.animate(withDuration: 0.6, animations: {
            self.plus.transform = CGAffineTransform(rotationAngle: -.pi/2)
            btnCloseAnimate(withFrame: CGRect(x: 273, y: 700, width: 71, height: 100), delay: 0.1, btnView: self.more)
            btnCloseAnimate(withFrame: CGRect(x: 152, y: 700, width: 71, height: 100), delay: 0.15, btnView: self.comment)
            btnCloseAnimate(withFrame: CGRect(x: 31, y: 700, width: 71, height: 100), delay: 0.2, btnView: self.sign)
            btnCloseAnimate(withFrame: CGRect(x: 273, y: 700, width: 71, height: 100), delay: 0.25, btnView: self.camera)
            btnCloseAnimate(withFrame: CGRect(x: 152, y: 700, width: 71, height: 100), delay: 0.3, btnView: self.ablum)
            btnCloseAnimate(withFrame: CGRect(x: 31, y: 700, width: 71, height: 100), delay: 0.35, btnView: self.text)
        }) { finished in
            self.blurView.isHidden = true
        }
    }
}
